relay-server/src/services/processor.rs

```diff
@@ -1214,184 +1234,191 @@ impl EnvelopeProcessorService {
-            event::store(state, &self.inner.config)?;
-            self.enforce_quotas(state)?;
+            state = event::store(state, &self.inner.config)?;
+            state = self.enforce_quotas(state)?.inner();
         });
 
         if state.has_event() {
-            event::scrub(state)?;
-            event::serialize(state)?;
+            state = event::scrub(state)?;
+            state = event::serialize(state)?;
         }
 
-        attachment::scrub(state);
+        state = attachment::scrub(state);
 
-        Ok(())
+        Ok(state)
     }
 
     /// Processes only transactions and transaction-related items.
-    fn process_transactions(
-        &self,
-        state: &mut ProcessEnvelopeState<TransactionGroup>,
-    ) -> Result<(), ProcessingError> {
-        profile::filter(state);
-        event::extract(state, &self.inner.config)?;
-        profile::transfer_id(state);
+    fn process_transactions<'a>(
+        &'_ self,
+        mut state: ProcessEnvelopeState<'a, TransactionGroup>,
+    ) -> Result<ProcessEnvelopeState<'a, TransactionGroup>, ProcessError<'a, TransactionGroup>>
+    {
+        state = profile::filter(state);
+        state = event::extract(state, &self.inner.config)?;
+        state = profile::transfer_id(state);
 
         if_processing!(self.inner.config, {
-            attachment::create_placeholders(state);
+            state = attachment::create_placeholders(state);
         });
 
-        event::finalize(state, &self.inner.config)?;
-        self.light_normalize_event(state)?;
-        dynamic_sampling::normalize(state);
-        event::filter(state)?;
-        dynamic_sampling::run(state, &self.inner.config);
+        state = event::finalize(state, &self.inner.config)?;
+        state = self.light_normalize_event(state)?;
+        state = dynamic_sampling::normalize(state);
+        state = event::filter(state)?;
+        state = dynamic_sampling::run(state, &self.inner.config);
 
         // We avoid extracting metrics if we are not sampling the event while in non-processing
         // relays, in order to synchronize rate limits on indexed and processed transactions.
         if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
-            self.extract_metrics(state)?;
+            state = self.extract_metrics(state)?;
         }
 
-        dynamic_sampling::sample_envelope_items(
+        state = dynamic_sampling::sample_envelope_items(
             state,
             &self.inner.config,
             &self.inner.global_config.current(),
         );
 
         if_processing!(self.inner.config, {
-            event::store(state, &self.inner.config)?;
-            self.enforce_quotas(state)?;
-            profile::process(state, &self.inner.config);
+            state = event::store(state, &self.inner.config)?;
+            let enforced_state = self.enforce_quotas(state)?;
+            state = profile::process(enforced_state, &self.inner.config);
```

Comment:
It seems wrong to go back from `EnforcedQuotasState` to ProcessEnvelopeState` here. Since we care about maintaining order, maybe `enforce_quotas` is a bad place to start, because it is only called in processing relays. Something like `event::filter` with a resulting `FilteredState` would be easier to enforce later on.

Reply:
Yeah, the enforced quotas type might be a bad example to start with